src/state/app_block_tree_view.rs

```diff
@@ -0,0 +1,91 @@
+/*
+    Copyright © 2023, ParallelChain Lab
+    Licensed under the Apache License, Version 2.0: http://www.apache.org/licenses/LICENSE-2.0
+*/
+
+use crate::hotstuff::types::QuorumCertificate;
+use crate::types::{
+    basic::{AppStateUpdates, BlockHeight, CryptoHash, Data, DataLen, Datum},
+    block::Block,
+    validators::ValidatorSet
+};
+
+use super::block_tree::BlockTreeError;
+use super::{block_tree::BlockTree, kv_store::KVStore};
+
+
+pub struct AppBlockTreeView<'a, K: KVStore> {
```

Comment:
I wonder if given the updated protocol this struct needs to be updated as well. Concretely, in the current protocol (and in the pipelined HotStuff protocol from the paper) a block doesn't necessarily get committed once it has great-grandchildren. Rather, it will get committed once it has 3 ancestors from consecutive views in the chain extending from it upwards. This may require arbitrarily many ancestors, depending on the protocol execution. So there might be more than 3 `AppStateUpdates` that have not been committed yet. @lyulka what do you think?

Reply:
Yes, we indeed should. The way this is currently implemented (in this PR) is incorrect. 

One way to correctly implement this is to walk back down the chain we're on until we hit the highest committed block, accumulating the pending `AppStateUpdate`s we encounter along the way.

Reply:
So instead of these 3 fields we'd have something like:
(1)
```rust
pub(super) ancestor_app_state_updates: Vec<AppStateUpdates>
```
or if an uncommitted block can have no associated pending app state updates (?):
(2)
```rust
pub(super) ancestor_app_state_updates: Vec<Option<AppStateUpdates>>
```
where the vector goes from parent block until highest committed block?

Reply:
Ok, I'll leave this decision for later. I implemented my suggested fix today morning, but I am open to changing it to what you are suggesting.

Reply:
Not going to implement now.